src/profiler.cpp

```diff
@@ -519,11 +520,49 @@ void Profiler::updateThreadName(jvmtiEnv* jvmti, JNIEnv* jni, jthread thread) {
         jvmtiThreadInfo thread_info;
         if (vm_thread != NULL && jvmti->GetThreadInfo(thread, &thread_info) == 0) {
             setThreadName(vm_thread->osThreadId(), thread_info.name);
+            addThreadToFilteredList(vm_thread->osThreadId(), thread_info.name);
             jvmti->Deallocate((unsigned char*)thread_info.name);
         }
     }
 }
 
+void Profiler::addThreadToFilteredList(int osThreadId, const char *name) {
+    MutexLocker lock(_filtered_tids_mutex);
+
+    if (_filter_threads == NULL) return;
+
+    if (strstr(name, _filter_threads)) {
+        _filtered_tids.push_back(osThreadId);
+    } else {
+        //might have matched before and needs to be removed after changing the name
+        _filtered_tids.erase(std::remove(_filtered_tids.begin(), _filtered_tids.end(), osThreadId),
+                             _filtered_tids.end());
+    }
+}
+
+void Profiler::removeFromFilteredList(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread) {
+    MutexLocker lock(_filtered_tids_mutex);
+    if (_threads && VMThread::available() && _filter_threads != NULL) {
+        VMThread *vm_thread = VMThread::fromJavaThread(jni, thread);
+        jvmtiThreadInfo thread_info;
+        if (vm_thread != NULL && jvmti->GetThreadInfo(thread, &thread_info) == 0) {
+            _filtered_tids.erase(std::remove(_filtered_tids.begin(), _filtered_tids.end(), vm_thread->osThreadId()),
+                                 _filtered_tids.end());
+            jvmti->Deallocate((unsigned char *) thread_info.name);
+        }
+    }
+}
+
+std::vector<int> Profiler::getFilteredTidsRange(int from, int count) {
+    std::vector<int> threads_for_this_tick;
```

Comment:
Why create a new vector each time?
Can't we walk over _filtered_tids?

Furthermore, with this approach the order of traversed tids will break if `_filtered_tids.size()` changes between iterations.